pkg/cmd/webhooks/forward.go

```diff
@@ -0,0 +1,253 @@
+package webhooks
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/gorilla/websocket"
+	"github.com/spf13/cobra"
+
+	"github.com/cli/cli/v2/api"
+	"github.com/cli/cli/v2/internal/config"
+	"github.com/cli/cli/v2/pkg/cmdutil"
+	"github.com/cli/cli/v2/pkg/iostreams"
+)
+
+const gitHubAPIProdURL = "http://api.github.com"
+
+var maxRetries = 3
+var webhookRcvServerURL = ""
+
+type hookOptions struct {
+	HttpClient func() (*http.Client, error)
+	IO         *iostreams.IOStreams
+	Config     func() (config.Config, error)
+
+	Host      string
+	EventType string
+	Repo      string
+	Port      int
+}
+
+func newCmdForward(f *cmdutil.Factory, runF func(*hookOptions) error) *cobra.Command {
+	opts := hookOptions{
+		HttpClient: f.HttpClient,
+		IO:         f.IOStreams,
+		Config:     f.Config,
+	}
+	cmd := cobra.Command{
+		Use:   "forward --event=<event_type> --repo=<repo> --port=<port> [--host=<host>]",
+		Short: "Receive test webhooks locally",
+		Example: heredoc.Doc(`
+			# create a dev webhook for the 'issue_open' event in the monalisa/smile repo in GitHub running locally, and
+			# forward payloads for the triggered event to localhost:9999
+
+			$ gh webhooks forward --event=issue_open --repo=monalisa/smile --port=9999 --host=api.github.localhost
+		`),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if opts.EventType == "" {
+				return cmdutil.FlagErrorf("`--event` flag required")
+			}
+			if opts.Repo == "" {
+				return cmdutil.FlagErrorf("`--repo` flag required")
+			}
+			if opts.Host == "" {
+				fmt.Fprintf(opts.IO.Out, "No --host specified, connecting to default GitHub.com \n")
+				opts.Host = gitHubAPIProdURL
+			}
+
+			if opts.Port == 0 {
+				fmt.Fprintf(opts.IO.Out, "No --port specified, printing webhook payloads to stdout \n")
+			}
+
+			if runF != nil {
+				return runF(&opts)
+			}
+
+			wsURLString, err := createHook(&opts)
+			if err != nil {
+				return err
+			}
+
+			wsURL, err := url.Parse(wsURLString)
+			if err != nil {
+				return err
+			}
+
+			cfg, err := opts.Config()
+			if err != nil {
+				return err
+			}
+			token, _ := cfg.AuthToken(opts.Host)
+
+			var retries int
+			events := make(chan wsEventReceived)
+			replies := make(chan httpEventForward)
+
+			go forwardEvents(&opts, events, replies)
+
+			for {
+				if retries >= maxRetries {
+					fmt.Fprintf(opts.IO.Out, "Unable to connect to webhooks server, forwarding stopped.\n")
+					return nil
+				}
+
+				done := make(chan struct{})
+				defer close(done)
+				conn, err := handleWebsocket(token, wsURL, &opts, events)
+
+				if err != nil {
+					unwrapped := errors.Unwrap(err)
+					var syscallErr *os.SyscallError
+					// If the error is a TCP handleWebsocket error, retry connecting
+					if errors.As(unwrapped, &syscallErr) && (unwrapped.Error() == "connect: connection refused") {
```

Comment:
I think if we can't connect to the websocket server, then something's pretty wrong like GitHub's GLB is down or something. So maybe we don't need to check and retry and just exit out. But up to you 👍🏼

Reply:
Would be good to retry just for network blips, they happen all the time. Since the logic is already there we can potentially extend the duration or number of retries to also account for things like the WS server being deployed or something like that.